portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -49,168 +49,203 @@
         blogMessagesReplies.add(message);
-        BlogMessages.remove(message);
+        blogMessages.remove(message);
       }
     }
-
-    for (BlogMessage post : BlogMessages) {
+ 
+    for (BlogMessage post : blogMessages) {
       for (BlogMessage reply : blogMessagesReplies) {
         if (reply.getParentID() == post.getTimestamp()) {
           post.addReply(reply);
         }
-      }
+      } 
     }
-
-    BlogHashMap blogMap = new BlogHashMap();
-    blogMap.putInMap(BlogMessages);
-
-    // Load messages from BlogHashMap and respond with gson.
-    LinkedList<BlogMessage> loadedBlogMessages = blogMap.getMessages(
-        tagsToSearch, numberOfCommentsToDisplay);
-
+ 
+    // Get user email
+    UserService userService = UserServiceFactory.getUserService();
+    String email = (String) userService.getCurrentUser().getEmail();
+ 
+    // Check to see if user follows any tags
+    if (LoadFollowedTags.hasFollowedTags(email)) {
+      List<String> newTags = LoadFollowedTags.getFollowedTags(email);
+      for (String tag : newTags) {
+        if (!tagsToSearch.contains(tag)) {
+          tagsToSearch.add(tag);
+        }
+      } 
+    }
+ 
     Gson gson = new Gson();
     response.setContentType("application/json;");
-    
-    response.getWriter().println(gson.toJson(loadedBlogMessages));
-    return;
+ 
+    response.getWriter().println(gson.toJson(sortAndLoadFromBlogHashMap(
+       blogMessages, tagsToSearch, numberOfCommentsToDisplay)));
   }
-    
+ 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get Post parameters.
+    // We can't process an empty message.
+    // Also a post without a tag will be assigned the default tag.
     String message = request.getParameter(MESSAGE_PARAMETER);
+    if (message == null || message.isEmpty()) {
+      response.setContentType("text/html");
+      response.getWriter().println("<h2> Cannot post empty message.</h2>");
+      return;
+    }
+ 
     String nickname = request.getParameter(SENDER_PARAMETER);
     String postTag = request.getParameter(TAG_PARAMETER);
-    long parentID = PARENT_ID_PARAMETER;
     if (postTag == null || postTag.isEmpty()) {
       postTag = InternalTags.defaultTag();
     }
-
+ 
+    long parentID = 0;
+ 
     // TODO: Handle replies later.
     List<BlogMessage> messageReplies = new ArrayList<BlogMessage>(); 
-  
-    //TODO: Handle image file sent with FormData.
-      
+ 
+    // TODO: Handle image file sent with FormData.
+ 
+    putBlogsInDatastore(postTag, message, nickname, messageReplies, parentID);
+ 
+    // Respond with the recent post.
+    // |LoadAllBlogsOrLast| returns the recent post if false is passed.
+    Gson gson = new Gson();
+ 
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(LoadAllBlogsOrLast(false)));   
+  }
+ 
+  // HELPER FUNCTION:.........................................................................
```

Comment:
The comment line may not be necessary? In some sense, all non-public methods are helper methods.

Reply:
It would be more useful to give some high-level description about what this method does, as well as description on non-trivial arguments.